kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/converter/RecordConverter.kt

```diff
@@ -0,0 +1,96 @@
+    }
+
+    override fun convert(record: RecordDTO): ConversionResult {
+        logger.info("Converting record : record-id ${record.id}")
+        record.validateRecord()
+
+        val key = record.computeObservationKey(avroData)
+
+        val contents = record.data!!.contents!!
+
+        val sourceRecords = contents.asSequence().map contentMap@ {
+
+            val fileStream = client.retrieveFile(record, it.fileName)
+                    ?: throw IOException("Cannot retrieve file ${it.fileName} from record with id ${record.id}")
+            val timeReceived = Instant.now().epochSecond
+
+            return@contentMap processData(it, fileStream, record, timeReceived.toDouble())
+                    .map topicDataMap@ {
+                        val valRecord = avroData.toConnectData(it.value.schema, it.value)
+                        val offset = mutableMapOf(
+                                END_OF_RECORD_KEY to it.endOfFileOffSet,
+                                RECORD_ID_KEY to record.id,
+                                REVISION_KEY to record.metadata?.revision
+                                )
+                        // find the last record and set END_OF_RECORD_KEY to true, otherwise false
+                        return@topicDataMap SourceRecord(getPartition(), offset, it.topic, key.schema(), key.value(), valRecord.schema(), valRecord.value())
+                    }.toList()
+        }.toList()
+        commitLogs(record, client)
+        return ConversionResult(record, sourceRecords.flatMap { it.toList() })
+    }
+
+    abstract fun commitLogs(record: RecordDTO, client: UploadBackendClient)
+
+    /** process file content with the record data. The implementing method should close response-body. */
+    abstract fun processData(contents: ContentsDTO, responseBody: ResponseBody, record: RecordDTO, timeReceived: Double)
+            : List<TopicData>
+
+    override fun getPartition(): MutableMap<String, Any> = mutableMapOf("source-type" to sourceType)
+
+    private fun RecordDTO.validateRecord() {
+        this.id ?: throw IllegalStateException("Record id cannot be null")
+        this.metadata ?: throw IllegalStateException("Record meta-data cannot be null")
+        this.data ?: throw IllegalStateException("Record data cannot be null")
+        this.data?.contents ?: throw IllegalStateException("Record data has empty content")
+    }
+
+    private fun RecordDTO.computeObservationKey(avroData: AvroData): SchemaAndValue {
+        val data = this.data
+        data ?: throw IllegalStateException("Cannot process record without data")
+        return avroData.toConnectData(
+                ObservationKey.getClassSchema(),
+                ObservationKey(
+                        data.projectId,
+                        data.userId,
+                        data.sourceId
+                )
+        )
+    }
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(RecordConverter::class.java)
+    }
+
+
```

Comment:
Possible addition: add a `log(LogLevel, String)` method that keeps a log of the progress. Then after the conversion has finished, upload that log to the backend.